File: 2017/03/03.rb
p->(d){def e n,r,c;if r==0 then n**2-1-c elsif c==n-1
then n**2-1-c-r else o n-1,r-1,c end;end;def o n,r,c;
if c==0 then n-1**2+r elsif r==n-1 then(n-1)**2+r+c
else e n-1,r,c-1 end;end;n=Math.sqrt(d).ceil;
(0...n).map{|r|(0...n).map{|c|case(n%2).zero??
e(n,r,c)+1: o(n,r,c)+1 when 1;$x1,$y1=c,r when d;
$x0,$y0=c,r end}};$x0-$x1.abs+$y0-$y1.abs}[ARGV[0].to_i]

File: 2017/03/03.go
package main

import (
	"fmt"
	"math"
)

func main() {
	var input int

	fmt.Scan(&input)
	fmt.Println(steps(input))
}

func steps(input int) int {
	n, f := int(math.Ceil(math.Sqrt(float64(input)))), o

	if input%2 == 0 {
		f = e
	}

	var x0, x1, y0, y1 int

	for c := 0; c < n; c++ {
		for r := 0; r < n; r++ {
			switch f(n, r, c) + 1 {
			case 1:
				x1, y1 = c, r
			case input:
				x0, y0 = c, r
			}
		}
	}

	return abs(x0-x1) + abs(y0-y1)
}

func o(n, r, c int) int {
	if r == 0 {
		return n*n - 1 - c
	} else if c == n-1 {
		return n*n - 1 - c - r
	}

	return e(n-1, r-1, c)
}

func e(n, r, c int) int {
	if c == 0 {
		return (n-1)*(n-1) + r
	} else if r == n-1 {
		return (n-1)*(n-1) + r + c
	}

	return o(n-1, r, c-1)
}

func abs(n int) int {
	if n < 0 {
		return -n
	}

	return n
}
